libraries/AP_HAL/GPIO.h

```diff
@@ -37,9 +32,25 @@ class AP_HAL::GPIO {
     /* Alternative interface: */
     virtual AP_HAL::DigitalSource* channel(uint16_t n) = 0;
 
+    enum INTERRUPT_TRIGGER_TYPE {
+        INTERRUPT_NONE,
+        INTERRUPT_FALLING,
+        INTERRUPT_RISING,
+        INTERRUPT_BOTH,
+    };
+
     /* Interrupt interface: */
-    virtual bool    attach_interrupt(uint8_t interrupt_num, AP_HAL::Proc p,
-            uint8_t mode) = 0;
+    //                                ret , pin    , state,timestamp
+    FUNCTOR_TYPEDEF(irq_handler_fn_t, void, uint8_t, bool, uint32_t);
+    virtual bool    attach_interrupt(uint8_t pin,
+                                     irq_handler_fn_t fn,
+                                     INTERRUPT_TRIGGER_TYPE mode) = 0;
```

Comment:
We could make this return false by default as well, since only two HALs implement it.